enhancements/project-request-limit.md

```diff
@@ -0,0 +1,293 @@
+* Some customers use the `ProjectRequestLimit` plugin in 3.x and thus are blocked on their path to OpenShift 4.x. Enabling
+ it on 4.x will unblock these customers.
+* The `kube-apiserver` is designed to be extended using validating/mutating admission webhook(s), we have the
+technology to easily build one, and we have the ability to create a simple operator to manage it.
+* The cluster admin can optionally enable the validating webhook on a cluster and configure limits via a custom
+resource. This will sever the "legacy" link between the plugin and the `kube-apiserver`.
+* We can provide the customer with seamless install and automatic upgrades by shipping the operator via OperatorHub 
+  independent of OpenShift release cycle.   
+
+
+### Goals
+1. Enable the `requestlimit.project.openshift.io/ProjectRequestLimit` admission plugin.
+2. Use existing extension points, libraries, and installation mechanisms in the manner we would recommend to external teams.
+3. Have a fairly straightforward way to install and enable this admission plugin.
+4. `rebootstrapping` must be possible.
+
+### Non-Goals
+1. Revisit how `requestlimit.project.openshift.io/ProjectRequestLimit` works. We want to lift it as is.
+2. Couple a slow moving admission plugin to a fast moving `kube-apiserver`.
+
+### Open Questions
+1. How do we productize the operator? Since this operator is not part of OpenShift core, it can essentially be built and
+ pushed to OperatorHub independently of OpenShift release cycle.
+3. How do we describe version skew limitations to OLM so our operator gets uninstalled *before* an illegal downgrade or
+ upgrade? This is a concrete case of the API we want to use isn't available before 1.16 and after 1.18, the previous API could be gone.
+
+## Proposal
+1. Create a `Validating` admission webhook server that provides `requestlimit.project.openshift.io/ProjectRequestLimit`.
+2. The admission webhook will enable `localhost` serving and `kube-apiserver` will talk to it using a `localhost` URL on the host network.
+2. Create an operator that manages all lifecycle aspects of the admission webhook via a custom resource.
+3. Productize the operator and ship it as a `RedHat operator` via `OperatorHub`. 
+4. Provide official documentation on how to interact with the `ProjectRequestLimit` operator.
+
+### User Stories [optional]
+
+*Story 1*: As an engineer I want to port `ProjectRequestLimit` plugin from 3.11 into a `Validating` admission webhook. 
+
+Notes:
+* Use  [generic-admission-server](https://github.com/openshift/generic-admission-server) to wire the validating admission webhook server.
+* The source code is here: https://github.com/openshift/origin/blob/release-3.11/pkg/project/apiserver/admission/requestlimit/admission.go 
+
+*Story 2*: As an engineer I want to write an operator that manages the `ProjectRequestLimit` admission webhook server.
+
+*Story 3*: As an engineer I want to ship `ProjectRequestLimit` as a RedHat operator via OperatorHub.
```

Comment:
all these are Developer Stories :) Or User Stories for OLM.

Reply:
okay, I rewrote the stories from the perspectives of the real user persona - `cluster admin`. Basically, we have four tasks: write the webhook, write the operator, productize and user documentation.